languages/rust/oso/src/host/to_polar.rs

```diff
@@ -18,25 +21,94 @@ use super::Host;
 ///
 /// For non-primitive types, the instance will be stored
 /// on the provided `Host`.
-pub trait ToPolar {
-    fn to_polar_value(&self, host: &mut Host) -> Value;
+/// ## Trait bounds
+///
+/// `ToPolar` requires types to be `Send + Sync`, since it
+/// is possible to store a `ToPolar` value on an `Oso` instance
+/// which can be shared between threads.
+///
+/// `ToPolar` implementors must also be concrete, sized types without
+/// any borrows.
+pub trait ToPolar: Send + Sync + Sized + 'static {
+    fn to_polar_value(self, host: &mut Host) -> Value {
+        let instance = Instance::new(self);
+        let instance = host.cache_instance(instance, None);
+        Value::ExternalInstance(ExternalInstance {
+            constructor: None,
+            repr: Some(std::any::type_name::<Self>().to_owned()),
+            instance_id: instance,
+        })
+    }
 
-    fn to_polar(&self, host: &mut Host) -> Term {
+    fn to_polar(self, host: &mut Host) -> Term {
         Term::new_from_ffi(self.to_polar_value(host))
     }
 }
 
+impl<C: crate::PolarClass + Send + Sync> ToPolar for C {
+    fn to_polar_value(self, host: &mut Host) -> Value {
+        let instance = Instance::new(self);
+        let instance = host.cache_instance(instance, None);
+        if host.get_class_from_type::<Self>().is_err() {
+            let class = Self::get_polar_class();
+            let name = Symbol(class.name.clone());
+            tracing::info!("class {} not previously registered, doing so now", name.0);
+            if let Err(e) = host.cache_class(class, name.clone()) {
+                tracing::debug!({ name = %name.0 }, "failed to cache class: {}", e);
```

Comment:
Let's add a TODO here, I think this shouldn't just be swallowed.

Reply:
Hmm, yeah currently this would happen if you register a class called "Foo" and then call `to_polar` on a _distinct_ class which is also called "Foo". I think blowing up in this case is probably better.